Cython/Utility/ObjectHandling.c

```diff
@@ -2873,3 +2914,52 @@ static CYTHON_INLINE PyObject *__Pyx_PyUnicode_ConcatInPlaceImpl(PyObject **p_le
     PyNumber_Add(a, b) : __Pyx_PyStr_Concat(a, b))
 #define __Pyx_PyStr_ConcatInPlaceSafe(a, b) ((unlikely((a) == Py_None) || unlikely((b) == Py_None)) ? \
     PyNumber_InPlaceAdd(a, b) : __Pyx_PyStr_ConcatInPlace(a, b))
+
+/////////////// FormatTypeName.proto ///////////////
+
+#if CYTHON_COMPILING_IN_LIMITED_API
+typedef PyObject *__Pyx_TypeName;
+#define __Pyx_FMT_TYPENAME "%U"
+static __Pyx_TypeName
+__Pyx_PyType_GetName(PyTypeObject* tp);
+#define __Pyx_DECREF_TypeName(obj) Py_XDECREF(obj)
+#else
+typedef const char *__Pyx_TypeName;
+#define __Pyx_FMT_TYPENAME "%.200s"
+#define __Pyx_PyType_GetName(tp) ((tp)->tp_name)
+#define __Pyx_DECREF_TypeName(obj)
+#endif
+
+/////////////// FormatTypeName ///////////////
+
+#if CYTHON_COMPILING_IN_LIMITED_API
+static __Pyx_TypeName
+__Pyx_PyType_GetName(PyTypeObject* tp)
+{
+    PyObject *name = __Pyx_PyObject_GetAttrStr((PyObject *)tp,
+                                               PYIDENT("__name__"));
+    if (unlikely(name == NULL) || unlikely(!PyUnicode_Check(name))) {
+        PyErr_Clear();
+        name = PYIDENT("?");
+        Py_INCREF(name);
+    }
+    return name;
+}
+#endif
+
+/////////////// RaiseUnexpectedTypeError.proto ///////////////
+
+static CYTHON_INLINE void __Pyx_RaiseUnexpectedTypeError(const char *expected,
+                                                         PyObject *obj);
+
+/////////////// RaiseUnexpectedTypeError ///////////////
+//@requires: FormatTypeName
+
+static CYTHON_INLINE void
+__Pyx_RaiseUnexpectedTypeError(const char *expected, PyObject *obj)
```

Comment:
`expected` should be a Python object here. It's almost certainly a name that is already part of the module's set of interned strings.

Reply:
Hmm this is a bit tricky to do from with the `PyrexTypes.py` caller. It seems there is no replacement of `PYIDENT()` in the output there, and there's also no code object around to call `get_py_string_const`. Would you think it's worth adding a `code` parameter to `type_test_code` just so I can call `get_py_string_const` there or would it be fine to stay with the `const char*`?

Reply:
Phew, right. `type_test_code()` is really stretching all limits here already.
Then let's keep it a `const char*`.